File: Day-7: Linked List and Arrays/2) Clone a Linked List with random and next pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/
class Solution {
    public Node copyRandomList(Node head) {
        Map<Node, Node> map = new HashMap<>();
        return clone(head, map);
    }
    public Node clone(Node head, Map<Node, Node> map)
    {
        if(head == null)
        {
            return null;
        }
        Node temp = new Node(head.val);
        map.put(head, temp);
        temp.next = clone(head.next, map);
        if(head.random != null)
        {
            temp.random = map.get(head.random);
        }
        return temp;
    }
}

File: Day-07: Linked List and Arrays/2) Clone a Linked List with random and next pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*, Node*> mp;
        Node *ans = solve(head, mp);
        return ans;
    }
    Node *solve(Node *head, unordered_map<Node*, Node*> &mp)
    {
        if(head == nullptr)
        {
            return nullptr;
        }
        Node *temp = new Node(head->val);
        mp[head] = temp;
        temp->next = solve(head->next, mp);
        if(head->random)
        {
            temp->random = mp[head->random];
        }
        return temp;
    }
};
